features/cellular/TARGET_GENERIC_MODEM/generic_modem_driver/ReferenceCellularInterface.cpp

```diff
@@ -0,0 +1,794 @@
+       if (success) {
+           return NSAPI_ERROR_OK;
+       }
+
+       return NSAPI_ERROR_AUTH_FAILURE;
+}
+
+static void set_nwk_reg_status_csd(unsigned int status)
+{
+    switch (status) {
+        case CSD_NOT_REGISTERED_NOT_SEARCHING:
+        case CSD_NOT_REGISTERED_SEARCHING:
+            break;
+        case CSD_REGISTERED:
+        case CSD_REGISTERED_ROAMING:
+            tr_debug("Registered for circuit switched service");
+            break;
+        case CSD_REGISTRATION_DENIED:
+            tr_debug("Circuit switched service denied");
+            break;
+        case CSD_UNKNOWN_COVERAGE:
+            tr_debug("Out of circuit switched service coverage");
+            break;
+        case CSD_SMS_ONLY:
+            tr_debug("SMS service only");
+            break;
+        case CSD_SMS_ONLY_ROAMING:
+            tr_debug("SMS service only");
+            break;
+        case CSD_CSFB_NOT_PREFERRED:
+            tr_debug("Registered for circuit switched service with CSFB not preferred");
+            break;
+        default:
+            tr_debug("Unknown circuit switched service registration status. %u", status);
+            break;
+    }
+
+    dev_info->reg_status_csd = static_cast<nwk_registration_status_csd>(status);
+}
+
+static void set_nwk_reg_status_psd(unsigned int status)
+{
+    switch (status) {
+        case PSD_NOT_REGISTERED_NOT_SEARCHING:
+        case PSD_NOT_REGISTERED_SEARCHING:
+            break;
+        case PSD_REGISTERED:
+        case PSD_REGISTERED_ROAMING:
+            tr_debug("Registered for packet switched service");
+            break;
+        case PSD_REGISTRATION_DENIED:
+            tr_debug("Packet switched service denied");
+            break;
+        case PSD_UNKNOWN_COVERAGE:
+            tr_debug("Out of packet switched service coverage");
+            break;
+        case PSD_EMERGENCY_SERVICES_ONLY:
+            tr_debug("Limited access for packet switched service. Emergency use only.");
+            break;
+        default:
+            tr_debug("Unknown packet switched service registration status. %u", status);
+            break;
+    }
+
+    dev_info->reg_status_psd = static_cast<nwk_registration_status_psd>(status);
+}
+
+static bool is_registered_csd()
+{
+  return (dev_info->reg_status_csd == CSD_REGISTERED) ||
+          (dev_info->reg_status_csd == CSD_REGISTERED_ROAMING) ||
+          (dev_info->reg_status_csd == CSD_CSFB_NOT_PREFERRED);
+}
+
+static bool is_registered_psd()
+{
+    return (dev_info->reg_status_psd == PSD_REGISTERED) ||
+            (dev_info->reg_status_psd == PSD_REGISTERED_ROAMING);
+}
+
+ReferenceCellularInterface::ReferenceCellularInterface(bool debugOn, PinName tx, PinName rx, int baud)
+{
+    _new_pin = NULL;
+    _pin = NULL;
+    _at = NULL;
+    _dcd = NULL;
+    _apn = "internet";
+    _uname = NULL;
+    _pwd = NULL;
+    _debug_trace_on = false;
+
+    // Set up File Handle
+    _fh = new BufferedSerial(tx, rx, BAUD_RATE);
```

Comment:
I don't see a compelling reason to change the current code, the argument that one should not allocate dynamic memory just for sake of traction doesn't make me thoroughly convinced . There are tools available for dynamic memory allocation tracking as well. As currently we have no idea how a USB driver will look like, I'm of the view that we should not touch the current code as the big picture is not very clear. Embedding filehandle into CellularInterface is not a good idea, so I request not to stress on this one.